### PR TITLE
Feature/recovery

### PR DESCRIPTION
**Pull request by IchBinTiMo**
implement data recovery to resolve data out of sync while the main node is dead.
implement the mechanism for the master node to distinguish the main node and replica node.

### CODE REVIEW

File: master_server/master_server.py [hunk 1]
```diff
@@ -77,6 +77,36 @@ def get_shard(state_abbreviation):
...
+def check_node_health_and_send_query(sharding_id, query):
+    node_pair = NODE_PAIRS[sharding_id]
+    main_node = node_pair[node_pair['main']]
+    try:
+        with socket.socket(socket.AF_INET,
+                            socket.SOCK_STREAM) as sharding_socket:
+            sharding_socket.connect(
+                (main_node['host'], main_node['port']))
+            print(
+                f"[Master Server] Sending CREATE query to Sharding {sharding_id + 1}"
```
Comment (Chengyen851010) [inline master_server/master_server.py]: Message要改成相對應的，這樣我們接著才好debug

File: master_server/master_server.py [hunk 2]
```diff
@@ -77,6 +77,36 @@ def get_shard(state_abbreviation):
     hash_value = hashlib.md5(state_abbreviation.encode()).hexdigest()
     return int(hash_value, 16) % 2  # Modulo 2 for two shards
 
+def check_node_health_and_send_query(sharding_id, query):
```
Comment (Chengyen851010) [inline master_server/master_server.py]:
```suggestion
def check_node_health_and_send_query(sharding_id, query, query_type): #quey_type = "CREATE", "SELECT"....
```

File: master_server/master_server.py [hunk 3]
```diff
@@ -77,6 +77,36 @@ def get_shard(state_abbreviation):
...
+            sharding_socket.sendall(query.encode('utf-8'))
+            response = sharding_socket.recv(1024)
+            print(
+                f"[Master Server] Received response from Sharding {sharding_id + 1}: {response.decode('utf-8')}"
+            )  # Troubleshooting print
+
+            return response
+    except Exception as e:
+        print(
+            f"[Master Server] Error connecting to Sharding {sharding_id + 1}: {e}"
```
Comment (Max851010) [inline master_server/master_server.py:100]:
```suggestion
            f"[Master Server] Error connecting to Sharding {sharding_id }: {e}"
```

File: master_server/master_server.py [hunk 4]
```diff
@@ -202,35 +232,38 @@ def handle_request(client_socket):
...
-                main_node = target_pair[target_pair['main']]
-
-                sharding_socket.connect(main_node['host'], main_node['port'])
-                sharding_socket.sendall(query.encode('utf-8'))
-                response = sharding_socket.recv(1024)
+
+            response = ""
+
+            for _ in range(2):
+                response = check_node_health_and_send_query(sharding_id, query)
```
Comment (Chengyen851010) [inline master_server/master_server.py]:
```suggestion
                response = check_node_health_and_send_query(sharding_id, query, "SELECT")
```

File: master_server/master_server.py [hunk 5]
```diff
@@ -202,35 +232,38 @@ def handle_request(client_socket):
...
-                    main_node = target_pair[target_pair['main']]
-
-                    sharding_socket.connect(main_node['host'],
-                                            main_node['port'])
-                    sharding_socket.sendall(query.encode('utf-8'))
-                    final_response += sharding_socket.recv(1024)
+                response = ""
+
+                for _ in range(2):
+                    response = check_node_health_and_send_query(i, query)
```
Comment (Chengyen851010) [inline master_server/master_server.py]:
```suggestion
                    response = check_node_health_and_send_query(i, query, "SELECT")
```

File: master_server/master_server.py [hunk 6]
```diff
@@ -246,47 +279,39 @@ def handle_request(client_socket):
...
-            target_pair = NODE_PAIRS[sharding_id]
-            main_node = target_pair[target_pair['main']]
 
-            sharding_socket.connect(main_node['host'], main_node['port'])
-            sharding_socket.sendall(query.encode('utf-8'))
-            response = sharding_socket.recv(1024)
+        response = ""
+
+        for _ in range(2):
+            response = check_node_health_and_send_query(sharding_id, query)
```
Comment (Chengyen851010) [inline master_server/master_server.py]:
```suggestion
            response = check_node_health_and_send_query(sharding_id, query, "INSERT")
```

File: master_server/master_server.py [hunk 7]
```diff
@@ -246,47 +279,39 @@ def handle_request(client_socket):
...
-                        sharding_response = sharding_socket.recv(1024)
-                        print(
-                            f"[Master Server] Received response from Sharding {sharding_id + 1}: {sharding_response.decode('utf-8')}"
-                        )
-                except Exception as e:
+            for sharding_id in range(2):
+                response = ""
+
+                for _ in range(2):
+                    response = check_node_health_and_send_query(sharding_id, query)
```
Comment (Chengyen851010) [inline master_server/master_server.py]:
```suggestion
                    response = check_node_health_and_send_query(sharding_id, query, "CREATE")
```

File: master_server/master_server.py [hunk 8]
```diff
@@ -318,6 +348,13 @@ def run_server():
     poller = select.poll()
     poller.register(server_socket, select.POLLIN)
 
+    # Start the node health manager
+    # node_health_manager = threading.Thread(
+    #     target=node_health_monitor,
+    # )
+
+    # node_health_manager.start()
+
```
Comment (Chengyen851010) [inline master_server/master_server.py]:
```suggestion
```

File: master_server/master_server.py [hunk 9]
```diff
@@ -300,15 +325,20 @@ def handle_request(client_socket):
...
         print("[Master Server] Closing server socket and releasing port...")
         server_socket.close()  # Close the server socket to release the port
 
+    # if node_health_manager:
+    #     node_health_manager.join()  # Wait for the node health manager to finish
+
 
 def run_server():
     global shutdown_flag
+    # global shutdown_flag, node_health_manager
```
Comment (Chengyen851010) [inline master_server/master_server.py]:
```suggestion
```

File: master_server/master_server.py [hunk 10]
```diff
@@ -300,15 +325,20 @@ def handle_request(client_socket):
...
+    # global shutdown_flag, server_socket, node_health_manager
     print("[Master Server] Received SIGINT. Shutting down...")
     shutdown_flag = True
     if server_socket:
         print("[Master Server] Closing server socket and releasing port...")
         server_socket.close()  # Close the server socket to release the port
 
+    # if node_health_manager:
+    #     node_health_manager.join()  # Wait for the node health manager to finish
+
```
Comment (Chengyen851010) [inline master_server/master_server.py]:
```suggestion
```

File: master_server/master_server.py [hunk 11]
```diff
@@ -300,15 +325,20 @@ def handle_request(client_socket):
 
 def handle_sigint(signum, frame):
     global shutdown_flag, server_socket
+    # global shutdown_flag, server_socket, node_health_manager
```
Comment (Chengyen851010) [inline master_server/master_server.py]:
```suggestion
```

File: master_server/master_server.py (hunk 1 above)
Comment (IchBinTiMo) [reply master_server/master_server.py]: fixed

**Comment by Chengyen851010** [comment]
Main server is all good. Can merge after approval of @shenyansyun and @kkkkenkkk44

File: master_server/master_server.py (hunk 2 above)
Comment (IchBinTiMo) [reply master_server/master_server.py]: fixed

File: master_server/master_server.py (hunk 4 above)
Comment (IchBinTiMo) [reply master_server/master_server.py]: fixed

File: master_server/master_server.py (hunk 5 above)
Comment (IchBinTiMo) [reply master_server/master_server.py]: fixed

File: master_server/master_server.py (hunk 6 above)
Comment (IchBinTiMo) [reply master_server/master_server.py]: fixed

File: master_server/master_server.py (hunk 7 above)
Comment (IchBinTiMo) [reply master_server/master_server.py]: fixed

File: master_server/master_server.py (hunk 8 above)
Comment (IchBinTiMo) [reply master_server/master_server.py]: fixed

File: master_server/master_server.py (hunk 9 above)
Comment (IchBinTiMo) [reply master_server/master_server.py]: fixed

File: master_server/master_server.py (hunk 10 above)
Comment (IchBinTiMo) [reply master_server/master_server.py]: fixed

File: master_server/master_server.py (hunk 11 above)
Comment (IchBinTiMo) [reply master_server/master_server.py]: fixed

**Review by shenyansyun — APPROVED** [review]

**Review by Chengyen851010 — APPROVED** [review]

**Review by kkkkenkkk44 — APPROVED** [review]

**Pull request merged**